src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PotentialEditorConfigDataSource.cs

```diff
@@ -0,0 +1,90 @@
+using System.ComponentModel.Composition;
+using System.Threading.Tasks;
+using System.Threading.Tasks.Dataflow;
+using Microsoft.VisualStudio.ProjectSystem.Properties;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS
+{
+    [Export(typeof(IFileWatchProviderDataSource))]
+    internal class PotentialEditorConfigDataSource : ProjectValueDataSourceBase<IFileWatchProvider>, IFileWatchProviderDataSource
+    {
+        private readonly ConfiguredProject _project;
+
+        private int _version;
+        private IDisposable _projectRuleSourceLink;
+
+        /// <summary>
+        /// The block that receives updates from the active tree provider.
+        /// </summary>
+        private IBroadcastBlock<IProjectVersionedValue<IFileWatchProvider>> _broadcastBlock;
+
+        /// <summary>
+        /// The public facade for the broadcast block.
+        /// </summary>
+        private IReceivableSourceBlock<IProjectVersionedValue<IFileWatchProvider>> _publicBlock;
+
+        [ImportingConstructor]
+        public PotentialEditorConfigDataSource(ConfiguredProject project)
+            : base(project.Services)
+        {
+            _project = project;
+        }
+
+        public override NamedIdentity DataSourceKey { get; } = new NamedIdentity(nameof(PotentialEditorConfigDataSource));
+
+        public override IComparable DataSourceVersion => _version;
+
+        public override IReceivableSourceBlock<IProjectVersionedValue<IFileWatchProvider>> SourceBlock => _publicBlock;
+
+        public FileChangeKinds ChangeKinds => FileChangeKinds.Added | FileChangeKinds.Removed;
+
+        protected override void Initialize()
+        {
+            base.Initialize();
+
+            _projectRuleSourceLink = _project.Services.ProjectSubscription.ProjectRuleSource.SourceBlock.LinkToAsyncAction(ProcessProjectChanged, ruleNames: PotentialEditorConfigFiles.SchemaName);
+
+            _broadcastBlock = DataflowBlockSlim.CreateBroadcastBlock<IProjectVersionedValue<IFileWatchProvider>>(nameFormat: nameof(PotentialEditorConfigDataSource) + "Broadcast {1}");
+            _publicBlock = _broadcastBlock.SafePublicize();
+        }
+
+        private async Task ProcessProjectChanged(IProjectVersionedValue<IProjectSubscriptionUpdate> projectSnapshot)
+        {
+            if (projectSnapshot.Value.CurrentState.TryGetValue(PotentialEditorConfigFiles.SchemaName, out IProjectRuleSnapshot ruleSnapshot))
+            {
+                var fileWatchProvider = new PotentialEditorConfigFileWatchProvider(this, ruleSnapshot.Items.Keys);
+                _version++;
+                ImmutableDictionary<NamedIdentity, IComparable> dataSources = ImmutableDictionary<NamedIdentity, IComparable>.Empty.Add(DataSourceKey, DataSourceVersion);
+
+                await _broadcastBlock.SendAsync(new ProjectVersionedValue<IFileWatchProvider>(fileWatchProvider, dataSources));
+            }
+        }
+
+        private class PotentialEditorConfigFileWatchProvider : IFileWatchProvider
```

Comment:
Nested classes go in separate files.

Reply:
No longer an issue; `IFileWatchProvider` has been replaced with a concrete `FileWatchProvider` at the CPS level.